### PR TITLE
interfaces/builtin: introduce raw-input interface

### PR DESCRIPTION
**Pull request by mardy**
Context: https://forum.snapcraft.io/t/what-interface-provides-access-to-dev-input/22657/11

You can use [this libinput snap](https://gitlab.com/mardy/snaps/-/tree/master/libinput/snap) to manually test this interface. The snap runs correctly, but the following denials are found in the audit log:
```
[104642.241275] audit: type=1400 audit(1620727863.291:2670): apparmor="DENIED" operation="open" profile="snap.libinput-tool.libinput-debug-events" name="/run/udev/data/+usb:3-4.1:1.0" pid=475087 comm="libinput-debug-" requested_mask="r" denied_mask="r" fsuid=0 ouid=0
[104642.241399] audit: type=1400 audit(1620727863.291:2671): apparmor="DENIED" operation="open" profile="snap.libinput-tool.libinput-debug-events" name="/run/udev/data/c189:265" pid=475087 comm="libinput-debug-" requested_mask="r" denied_mask="r" fsuid=0 ouid=0
[104642.241559] audit: type=1400 audit(1620727863.291:2672): apparmor="DENIED" operation="open" profile="snap.libinput-tool.libinput-debug-events" name="/run/udev/data/c189:257" pid=475087 comm="libinput-debug-" requested_mask="r" denied_mask="r" fsuid=0 ouid=0
[104642.241603] audit: type=1400 audit(1620727863.291:2673): apparmor="DENIED" operation="open" profile="snap.libinput-tool.libinput-debug-events" name="/run/udev/data/c189:256" pid=475087 comm="libinput-debug-" requested_mask="r" denied_mask="r" fsuid=0 ouid=0
[104642.241688] audit: type=1400 audit(1620727863.291:2674): apparmor="DENIED" operation="open" profile="snap.libinput-tool.libinput-debug-events" name="/run/udev/data/+pci:0000:0b:00.3" pid=475087 comm="libinput-debug-" requested_mask="r" denied_mask="r" fsuid=0 ouid=0
[104642.241753] audit: type=1400 audit(1620727863.291:2675): apparmor="DENIED" operation="open" profile="snap.libinput-tool.libinput-debug-events" name="/run/udev/data/+pci:0000:00:08.1" pid=475087 comm="libinput-debug-" requested_mask="r" denied_mask="r" fsuid=0 ouid=0
[104642.309204] audit: type=1400 audit(1620727863.359:2676): apparmor="DENIED" operation="open" profile="snap.libinput-tool.libinput-debug-events" name="/run/udev/data/+usb:3-4.1:1.1" pid=475087 comm="libinput-debug-" requested_mask="r" denied_mask="r" fsuid=0 ouid=0
[104642.309328] audit: type=1400 audit(1620727863.359:2677): apparmor="DENIED" operation="open" profile="snap.libinput-tool.libinput-debug-events" name="/run/udev/data/c189:265" pid=475087 comm="libinput-debug-" requested_mask="r" denied_mask="r" fsuid=0 ouid=0
[104642.309415] audit: type=1400 audit(1620727863.359:2678): apparmor="DENIED" operation="open" profile="snap.libinput-tool.libinput-debug-events" name="/run/udev/data/c189:257" pid=475087 comm="libinput-debug-" requested_mask="r" denied_mask="r" fsuid=0 ouid=0
[104642.309501] audit: type=1400 audit(1620727863.359:2679): apparmor="DENIED" operation="open" profile="snap.libinput-tool.libinput-debug-events" name="/run/udev/data/c189:256" pid=475087 comm="libinput-debug-" requested_mask="r" denied_mask="r" fsuid=0 ouid=0
```

### CODE REVIEW

File: interfaces/builtin/raw_input.go
Comment (bboozzoo) [inline interfaces/builtin/raw_input.go:55]: touch screen? I see this in x11, wayland and mir interfaces but no comments there.

File: interfaces/builtin/raw_input.go
Comment (bboozzoo) [inline interfaces/builtin/raw_input.go:48]: Some input devices may actually be from acpi (eg. LNXPWR which is a power button), or from platform (iirc gpios would lead there). Not sure we need to allow reading those right now. Maybe it's better to wait until someone complains.

**Review by bboozzoo — APPROVED** [review]
Name probably needs and ack from @pedronis

File: interfaces/builtin/raw_input.go
Comment (mardy) [reply interfaces/builtin/raw_input.go:48]: As a matter of facts, I see these messages in the audit log when running the libinput test application (in `--devmode`, that's why they appear as `ALLOWED`:
```
[99572.990328] audit: type=1400 audit(1620722794.063:2261): apparmor="ALLOWED" operation="open" profile="snap.libinput-tool.libinput-debug-events" name="/run/udev/data/+acpi:LNXPWRBN:00" pid=458021 comm="libinput-debug-" requested_mask="r" denied_mask="r" fsuid=0 ouid=0
[99572.990351] audit: type=1400 audit(1620722794.063:2262): apparmor="ALLOWED" operation="open" profile="snap.libinput-tool.libinput-debug-events" name="/run/udev/data/+acpi:LNXSYSTM:00" pid=458021 comm="libinput-debug-" requested_mask="r" denied_mask="r" fsuid=0 ouid=0
[99572.990671] audit: type=1400 audit(1620722794.063:2263): apparmor="ALLOWED" operation="open" profile="snap.libinput-tool.libinput-debug-events" name="/run/udev/data/+acpi:PNP0C0C:00" pid=458021 comm="libinput-debug-" requested_mask="r" denied_mask="r" fsuid=0 ouid=0
[99572.990695] audit: type=1400 audit(1620722794.063:2264): apparmor="ALLOWED" operation="open" profile="snap.libinput-tool.libinput-debug-events" name="/run/udev/data/+acpi:LNXSYBUS:00" pid=458021 comm="libinput-debug-" requested_mask="r" denied_mask="r" fsuid=0 ouid=0
[99572.990720] audit: type=1400 audit(1620722794.063:2265): apparmor="ALLOWED" operation="open" profile="snap.libinput-tool.libinput-debug-events" name="/run/udev/data/+acpi:LNXSYSTM:00" pid=458021 comm="libinput-debug-" requested_mask="r" denied_mask="r" fsuid=0 ouid=0
[99572.991140] audit: type=1400 audit(1620722794.063:2266): apparmor="ALLOWED" operation="open" profile="snap.libinput-tool.libinput-debug-events" name="/run/udev/data/+usb:1-10:1.0" pid=458021 comm="libinput-debug-" requested_mask="r" denied_mask="r" fsuid=0 ouid=0
[99572.991180] audit: type=1400 audit(1620722794.063:2267): apparmor="ALLOWED" operation="open" profile="snap.libinput-tool.libinput-debug-events" name="/run/udev/data/c189:3" pid=458021 comm="libinput-debug-" requested_mask="r" denied_mask="r" fsuid=0 ouid=0
```
I'm now trying to see if I can install the application with strict confinement, to see if these accesses are really needed.

File: interfaces/builtin/raw_input.go
Comment (mardy) [reply interfaces/builtin/raw_input.go:63]: Yep.

File: interfaces/builtin/raw_input.go
Comment (pedronis) [inline interfaces/builtin/raw_input.go:40]: raw_usb doesn't have bind but we are not very consistent about this apparently

File: interfaces/builtin/raw_input.go
Comment (pedronis) [inline interfaces/builtin/raw_input.go:54]: do we need this? raw_usb doesn't have this (but maybe it was supposed to be used together with something else)

**Review by pedronis — COMMENTED** [review]
thanks, overall looks good to me, a couple of wonderings about things this specifies that raw_usb doesn't. I will mark this for security review now

**Comment by pedronis** [comment]
this will need additions to the spread tests

File: interfaces/builtin/raw_input.go
Comment (mardy) [reply interfaces/builtin/raw_input.go:40]: This is needed, or libinput failes to initialize udev.

File: interfaces/builtin/raw_input.go
Comment (mardy) [reply interfaces/builtin/raw_input.go:54]: Same as above: libinput failes to initialize udev without this.

**Review by stolowski — APPROVED** [review]
Looks good code-wise. I can't comment on actual permissions, not familiar with raw input requirements.

**Comment by bboozzoo** [comment]
The spread test looks ok. However, our static checks complain about:
```
 Missing high-level test for interface 'raw-input'. Please add to:
* tests/lib/snaps/test-snapd-policy-app-consumer/meta/snap.yaml
* tests/lib/snaps/test-snapd-policy-app-provider-core/meta/snap.yaml (if needed)
* tests/lib/snaps/test-snapd-policy-app-provider-classic/meta/snap.yaml (if needed)
```

Please take a look at those tests, IIRC it's just a matter of adding the relevant plug for respective snaps.

File: interfaces/builtin/raw_input.go
Comment (alexmurray) [reply interfaces/builtin/raw_input.go:56]: c189 is for USB Serial Converters as per https://www.kernel.org/doc/Documentation/admin-guide/devices.txt so I don't think this is relevant for this interface (not sure why libinput is trying to read this but I think this can be safely ignored).

I also agree that for the ACPI related ones as it is not as clear whether these make sense for this interface, perhaps these should be ignored for now.

**Review by alexmurray — APPROVED** [review]
LGTM - from a security perspective this interface contains just the minimum required for an application which requires access to the input subsystem. Also the tests look comprehensive as well. Thanks :)

**Comment by pedronis** [comment]
marking myself to go over Alex's feedback and also consider if we should tweak some other interface for consistency

File: interfaces/builtin/raw_input.go
Comment (pedronis) [reply interfaces/builtin/raw_input.go:56]: @mardy 

> I'm now trying to see if I can install the application with strict confinement, to see if these accesses are really needed.

what was the outcome of that? I suppose those denials weren't blockers?

As other have mentioned I don't we want to add ACPI devices or c189 here

File: interfaces/builtin/raw_input.go
Comment (pedronis) [reply interfaces/builtin/raw_input.go:40]: for consistency we should make a separate PR that adds this also to raw-usb

File: interfaces/builtin/raw_input.go
Comment (pedronis) [reply interfaces/builtin/raw_input.go:54]: same about raw-usb

**Review by pedronis — COMMENTED** [review]
thank you. a question and otherwise a comment about making raw-usb consistent with what this does (but separately)

File: interfaces/builtin/raw_input.go
Comment (mardy) [reply interfaces/builtin/raw_input.go:56]: The result is that these accesses are not really needed, the application works fine without them.

**Review by pedronis — APPROVED** [review]
thank you

**Pull request merged**
